text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,115 @@
+
+The version value can be used to enforce a specific version to be retrieved, and otherwise fail if that specific version is not found. On the other hand, the latest version will be retrieved if the version value is omitted from the XOR-URL.
+
+### Paths
+Given that a referenced MutableData can effectively be the root NFS/Files container of the files of a safesite, any MD XOR-URL can also specify a path which needs to be resolved, just like when using the DNS/public-names system to resolve the path in a URL.
+
+### Browsable content
+Currently, when a public-name URL is resolved to an NFS/Files container which doesn’t have an `index.html` file, the browser simply shows an error stating that the safesite content was not found.
+
+For a MD XOR-URL that doesn’t have a path and there is no `index.html` file, the resolver can return the raw content of the MD (i.e. its key-value entries), and the browser can automatically generate an HTML page which makes the content browsable, generating links to other data when an entry’s key/value is a `safe://` string, in an analogous/similar way to how web servers on the current internet allow browsing on folders/directories.
+
+An specific MD entry key could be also supported (e.g. `__non_browsable`), that the owner of the MD can insert in the MD if the “browsable” feature should not be enabled. Although this cannot be really enforced but offered as a feature to be optionally supported by some clients like our SAFE browser.
+
+### XOR-URLs specification
+The following are the main requirements for the encoding to be used to generate the XOR-URLs:
+
+- Be able to support new and different types of base encodings and hash functions for the XOR addresses in the future.
+- Include the content type within the XOR URL which would allow the client app to correctly render the data to the user especially when referencing an ImmutableData.
+
+It is proposed the use of [multiformats](https://github.com/multiformats) and [CID](https://github.com/ipld/cid), which allow us to cover the above requirements. A CID identifier can be used in the XOR-URL for specifying the XOR address part, and have additional parts to support MutableData’s type tag, version, as well as the path and query parts. SAFE URLs can then be defined in the following way (BNF 1-like):
+```
+<safe-url>            := 'safe://' ( <xor-uri> | <public-name-uri> )
+
+<public-name-uri>     :=  [<service> '.'] <public-name> <path-query-fragment>
+
+<xor-uri>             := <immutable-data-uri> | <mutable-data-uri>
+
+<immutable-data-uri>  := <cid> <query-fragment>
+
+<mutable-data-uri>    := <cid> ':' <type-tag> ['+' <content-version>] <path-query-fragment>
+
+<path-query-fragment> := ['/' <path>] <query-fragment>
+
+<query-fragment>      := ['?' <query>] ['#' <fragment>]
+```
```

Comment:
I'm not sure what you mean by a `key` , are you by any chance thinking along the lines of what [they defined as human-readable CIDs](https://github.com/ipld/cid#human-readable-cids)? 
If so , we could consider having the browser to show a tooltip with the human-readable CID when hovering over a XOR URL in the address bar, if that happens to be useful, and/or have the browser to automatically convert a human-readable CID into the CID string when the user pasted/entered it in the address bar as it can be good for those who understand the format to fetch content at a any given xorname. I would say this can be all done but I doubt it should be part of the resolver itself but perhaps provided by the client app, e.g. in our case the browser.

As per having shorter CIDs like you suggested above @ustulation , it would be nice, but shortening it as you seem to imply would break the encoding of the CID string, so you'd need some type of encoding which can be shortened that way and still be able to get the (shortened) xor address out of it

Reply:
I meant MData Key from the key-value pair.

On shortning, IDK how it breaks it. So say in base x you get `hsh121difvoc..` etc., if you trimmed it to `hsh1` you could convert that back to repr in base 16 (which currently what the network understand) and use that portion. So it will give you are shorter XOR address like explained in the overall [comment here](https://github.com/maidsafe/rfcs/pull/281#pullrequestreview-174881487). But again, if that might be too much of a digression and could hijack comments directly (and more closely) related to this RFC

Reply:
For getting a specific MData key, then yes you could use the #fragment which is what it's the standard way of referring to a specific part of the document you are targeting with the URL, e.g. the WebID spec also makes use of the #fragment to refer to one of the subgraphs, so I wouldn't do that as part of the resolver's job, since the #fragment is usually used by the app consuming the content.

Shortening the xorname to 4 or 6 bytes wouldn't still give you a 4 or 6 bytes XOR-URL, remember the CID encodes some others bytes, so yes, it sounds like it can be done depending which encoding you use (e.g. CID in this case) you won't still be able to shorten it to 6 bytes but a few more, which could still be good I guess. 
Anyway I think there is a major concern with shortening it as you suggest, it could be breaking the immutability of the relationship between the XOR-URL and the target content, e.g. you give me a short XOR-URL, and I can fetch an immutable content since it can resolve to an unique xorname, until one day someone stores another piece of data (at a xorname which has the same initial bytes) which makes that XOR-URL un-resolvable as it doesn't resolve to an unique address, right?

Reply:
>  as it doesn't resolve to an unique address, right?
Ya that's true (it's same as git, as mentioned), but probably not worth the hassle as you say.

> For getting a specific MData key, then yes you could use the #fragment which is what it's the standard way of referring to a specific part of the document you are targeting with the URL
Would it be better if you exemplified it in the RFC ? Perhaps a few more examples to illustrate would give an easier understanding while reading the RFC. I would also propose to explain the deconstruction in examples if it's needed. Something like:
```
to do X: <abcd>:://<efgh>:<ijkl>?<mnop>#<qrst>
where abcd=cid
           efgh=blah
 ... etc.
```

Reply:
Agreed, I also think it will help to have some details of the decoding, I'll work on adding it @ustulation